### ISSUE TITLE
Add support for environment based configuration

### ISSUE THREAD

**Issue by tristanpemble** [datawire/forge#23]
When running my application deploys, a lot of how they are configured depends on the target environment. For example, running MySQL locally I might only want to allocate 1GB of memory, but in production it would be much higher.

One of the ways I was thinking this could be possible, is if the `service.yml` allowed Jinja2 templating in the configuration values. This could be a powerful feature, especially if it were combined with a Jinja2 filter that allowed you to pull an environment variable. Imagine this:

```yaml
# ./service.yml
name: my-microservice
memory: "{{ env('MEMORY') | default('0.5G') }}"
```

This could of course be handled in the K8S template, too:

```yaml
# ./k8s/deployment.yaml
# ...
         resources:
           limits:
             memory: {{ env('MEMORY') | default('0.5G') }}
             cpu: {{ env('CPU') | default('0.25') }}
# ...
```

But I think the self documenting nature of the first example, by being contained in service.yml, is more advantageous.

This would allow you to use a local `.env` file in development, with something like `source .env && forge deploy`. It would also be helpful for handling deploys in CI environments, without storing production values in the Git repository.

**Comment by tristanpemble** [comment]
Fat-fingered the submit button before finishing up the issue description, reopening now that I have written out the full feature request.

**Comment by tristanpemble** [comment]
Thinking about it more, I think just passing the environment in as a dict to the templates would be all that is really necessary:

```yaml
# ./service.yml
name: my-microservice
memory: 0.5G
cpu: 0.25
```

```yaml
# ./k8s/deployment.yaml
# ...
         resources:
           limits:
             memory: {{ env.MEMORY | default(service.memory) }}
             cpu: {{ env.CPU | default(service.cpu) }}
# ...
```

the downside being that there's no list of environment variables available directly in the `service.yml`, which can just be managed by documenting them.

The Jinja2 templating in `service.yml` might be useful in other contexts, but would be unnecessary here.

**Comment by rhs** [comment]
I implemented both the jinja2 templating of service.yaml and passing env into the k8s templates as you describe, so I *believe* either of your examples should work.

I'm currently working on adding some more docs. I'll post here again when the docs are done, but if you install the latest, you should be able to play around with the new features you've described.

**Comment by rhs** [comment]
I just released a forge 0.3.1 with revamped output based on this thread. I think I was able to capture the spirit of the multiplexed output discussed.

I didn't end up getting anything workable with scrollbars. Basically after playing with it for a bit, I came to the following way of thinking:

Pretty much everything forge does is mostly instantaneous (on a functioning network) with the exception of docker build and docker push, and both of those things have extremely high variance. A first time build or push can take many minutes if no layers are cached, or they can take a few seconds, and there isn't really a good way for me to predict this. A scrollbar is therefore very likely to get stuck in one spot and make it appear is if forge is hanging under some circumstances.

So to deal with this I ended up going with a fairly simple scheme where by default docker build and docker push output are relayed (multiplexed if necessary) to the user, followed by a success/error summary at the end. The  verbose option operates similarly except outputs all commands and http interactions, not just docker build & push.

I think this is definitely better than before, although I'm sure more usage will elicit more useful feedback. ;-) I'm gonna call this issue done, but please submit more issues if you have any thoughts on improving the new output.
